test/server/test_adapter.py

```diff
@@ -901,3 +922,262 @@ def get_counts():
+                    shown=unique_ids[0],
+                    endorsed=[clust1_leaves],
+                ),
+            )
+            # Can send redundant data
+            self.backend.insert_judgement(
+                judgement=Judgement(
+                    user_id=alice_id,
+                    shown=unique_ids[0],
+                    endorsed=[clust1_leaves],
+                ),
+            )
+            assert self.backend.clusters.count() == original_cluster_num
+
+            # Now split a cluster
+            clust2_leaves = get_leaf_ids(unique_ids[1])
+            self.backend.insert_judgement(
+                judgement=Judgement(
+                    user_id=alice_id,
+                    shown=unique_ids[1],
+                    endorsed=[clust2_leaves[:1], clust2_leaves[1:]],
+                ),
+            )
+            # Now, two new clusters should have been created
+            assert self.backend.clusters.count() == original_cluster_num + 2
+
+            # Let's check failures
+            # First, confirm that the following leaves don't exist
+            fake_leaves = [10000, 10001]
+            with pytest.raises(MatchboxDataNotFound):
+                self.backend.validate_ids(fake_leaves)
+            # Now, let's test an exception is raised
+            with pytest.raises(MatchboxDataNotFound):
+                self.backend.insert_judgement(
+                    judgement=Judgement(
+                        user_id=alice_id, shown=unique_ids[0], endorsed=[fake_leaves]
+                    ),
+                )
+
+            # Now, let's try to get the judgements back
+            # Data gets back in the right shape
+            judgements, expansion = self.backend.get_judgements()
+            judgements.schema.equals(SCHEMA_JUDGEMENTS)
+            expansion.schema.equals(SCHEMA_CLUSTER_EXPANSION)
+
+            # Only one user ID was used
+            assert judgements["user_id"].unique().to_pylist() == [alice_id]
+            # The first shown cluster is repeated because we judged it twice
+            # The second shown cluster is repeated because we split it (see above)
+            assert sorted(judgements["shown"].to_pylist()) == sorted(
+                [unique_ids[0], unique_ids[0], unique_ids[1], unique_ids[1]]
+            )
+            # On the other hand, the root-leaf mapping table has no duplicates
+            assert len(expansion) == 4  # 2 shown clusters + 2 new endorsed clusters
+
+            # Let's massage tables into a root-leaf dict for all endorsed clusters
+            endorsed_root, endorsed_leaves = (
+                pl.from_arrow(judgements)
+                .join(pl.from_arrow(expansion), left_on="endorsed", right_on="root")[
+                    ["endorsed", "leaves"]
+                ]
+                .to_dict(as_series=False)
+                .values()
+            )
+            endorsed_dict = dict(zip(endorsed_root, endorsed_leaves, strict=True))
```

Comment:
AI reckons this should work.

```suggestion
            endorsed_dict = dict(
                pl.from_arrow(judgements)
                .join(pl.from_arrow(expansion), left_on="endorsed", right_on="root")
                .select(["endorsed", "leaves"])
                .rows()
            )
```